components/browser/engine-gecko-beta/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -224,6 +229,69 @@ class GeckoEngineSession(
         }
     }
 
+    @Suppress("ComplexMethod")
+    internal fun createContentDelegate() = object : GeckoSession.ContentDelegate {
+        override fun onContextMenu(
+            session: GeckoSession,
+            screenX: Int,
+            screenY: Int,
+            uri: String?,
+            elementType: Int,
+            elementSrc: String?
+        ) {
+            when (elementType) {
+                ELEMENT_TYPE_AUDIO ->
+                    elementSrc?.let {
+                        notifyObservers { onLongPress(HitResult.AUDIO(it)) }
+                    }
+                GeckoSession.ContentDelegate.ELEMENT_TYPE_VIDEO ->
+                    elementSrc?.let {
+                        notifyObservers { onLongPress(HitResult.VIDEO(it)) }
+                    }
+                ELEMENT_TYPE_IMAGE -> {
+                    when {
+                        elementSrc != null && uri != null ->
+                            notifyObservers { onLongPress(HitResult.IMAGE_SRC(elementSrc, uri)) }
+                        elementSrc != null ->
+                            notifyObservers { onLongPress(HitResult.IMAGE(elementSrc)) }
+                    }
+                }
+                ELEMENT_TYPE_NONE -> {
+                    elementSrc?.let {
+                        when {
+                            it.isPhone() -> notifyObservers { onLongPress(HitResult.PHONE(it)) }
+                            it.isEmail() -> notifyObservers { onLongPress(HitResult.EMAIL(it)) }
+                            it.isGeoLocation() -> notifyObservers { onLongPress(HitResult.GEO(it)) }
+                            else -> notifyObservers { onLongPress(HitResult.UNKNOWN(it)) }
+                        }
+                    } ?: notifyObservers { onLongPress(HitResult.UNKNOWN("")) }
+                }
+            }
```

Comment:
It looks like all branches call `notifyObservers { onLongPress(...) }` so I wonder if you could make the `when` statement return the HitResult and then call notifyObservers ... with that.